src/add-text-track-data.js

```diff
@@ -66,19 +66,38 @@ const addTextTrackData = function(sourceHandler, captionArray, metadataArray) {
   if (metadataArray) {
     metadataArray.forEach(function(metadata) {
       let time = metadata.cueTime + this.timestampOffset;
+      let videoDuration = Number.isFinite(this.mediaSource_.duration) ?
+        this.mediaSource_.duration :
+        Number.MAX_VALUE;
 
       metadata.frames.forEach(function(frame) {
         let cue = new Cue(
-            time,
-            time,
-            frame.value || frame.url || frame.data || '');
+          time,
+          videoDuration,
```

Comment:
This will generate a `cuechange` event for each cue that is added before the `currentTime`. It might make more sense to start with zero-length cues and extend their durations instead of starting at full-length and shortening them below.